hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/replication/health/TestRatisReplicationCheckHandler.java

```diff
@@ -0,0 +1,560 @@
+        ReplicationManagerReport.HealthState.OVER_REPLICATED));
+  }
+
+  @Test
+  public void testOverReplicatedContainerFixedByPending() {
+    ContainerInfo container = createContainerInfo(repConfig);
+    Set<ContainerReplica> replicas =  createReplicas(container.containerID(),
+        Pair.of(IN_SERVICE, 0), Pair.of(IN_SERVICE, 0),
+        Pair.of(IN_SERVICE, 0), Pair.of(IN_SERVICE, 0));
+
+    List<ContainerReplicaOp> pending = new ArrayList<>();
+    pending.add(ContainerReplicaOp.create(
+        DELETE, MockDatanodeDetails.randomDatanodeDetails(), 0));
+
+    requestBuilder.setContainerReplicas(replicas)
+        .setPendingOps(pending)
+        .setContainerInfo(container);
+    OverReplicatedHealthResult result = (OverReplicatedHealthResult)
+        healthCheck.checkHealth(requestBuilder.build());
+    Assert.assertEquals(HealthState.OVER_REPLICATED, result.getHealthState());
+    Assert.assertEquals(1, result.getExcessRedundancy());
+    Assert.assertTrue(result.isSufficientlyReplicatedAfterPending());
+
+    Assert.assertTrue(healthCheck.handle(requestBuilder.build()));
+    Assert.assertEquals(0, repQueue.underReplicatedQueueSize());
+    // Fixed by pending so nothing queued.
+    Assert.assertEquals(0, repQueue.overReplicatedQueueSize());
+    // Still over replicated, so the report should contain it
+    Assert.assertEquals(1, report.getStat(
+        ReplicationManagerReport.HealthState.OVER_REPLICATED));
+  }
+
+  @Test
+  public void testOverReplicatedContainerWithMaintenance() {
+    ContainerInfo container = createContainerInfo(repConfig);
+    Set<ContainerReplica> replicas =  createReplicas(container.containerID(),
+        Pair.of(IN_SERVICE, 0), Pair.of(IN_SERVICE, 0),
+        Pair.of(IN_SERVICE, 0), Pair.of(IN_SERVICE, 0),
+        Pair.of(IN_MAINTENANCE, 0), Pair.of(DECOMMISSIONED, 0));
+
+    requestBuilder.setContainerReplicas(replicas)
+        .setContainerInfo(container);
+    OverReplicatedHealthResult result = (OverReplicatedHealthResult)
+        healthCheck.checkHealth(requestBuilder.build());
+    Assert.assertEquals(HealthState.OVER_REPLICATED, result.getHealthState());
+    Assert.assertEquals(1, result.getExcessRedundancy());
+    Assert.assertFalse(result.isSufficientlyReplicatedAfterPending());
+
+    Assert.assertTrue(healthCheck.handle(requestBuilder.build()));
+    Assert.assertEquals(0, repQueue.underReplicatedQueueSize());
+    Assert.assertEquals(1, repQueue.overReplicatedQueueSize());
+    Assert.assertEquals(1, report.getStat(
+        ReplicationManagerReport.HealthState.OVER_REPLICATED));
+  }
+
+  @Test
+  public void testOverReplicatedContainerDueToMaintenance() {
+    ContainerInfo container = createContainerInfo(repConfig);
+    Set<ContainerReplica> replicas = createReplicas(container.containerID(),
+        Pair.of(IN_SERVICE, 0), Pair.of(IN_SERVICE, 0),
+        Pair.of(IN_SERVICE, 0), Pair.of(IN_MAINTENANCE, 0),
+        Pair.of(IN_MAINTENANCE, 2));
```

Comment:
The replica index should be 0 instead of 2 I think. `(IN_MAINTENANCE, 0)`

Reply:
yes, well spotted.